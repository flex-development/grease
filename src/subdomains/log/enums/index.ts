/**
 * @file Entry Point - Enums
 * @module grease/log/enums
 */

export { default as LogLevel } from './log-level'
export { default as UserLogLevel } from './log-level-user'
export { default as LogType } from './log-type'
