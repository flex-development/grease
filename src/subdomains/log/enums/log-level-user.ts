/**
 * @file Enums - UserLogLevel
 * @module grease/log/enums/UserLogLevel
 */

import type LogLevel from './log-level'

/**
 * User-friendly log levels.
 *
 * @enum {Lowercase<keyof typeof LogLevel>}
 */
enum UserLogLevel {
  DEBUG = 'debug',
  ERROR = 'error',
  FATAL = 'fatal',
  INFO = 'info',
  LOG = 'log',
  SILENT = 'silent',
  TRACE = 'trace',
  VERBOSE = 'verbose',
  WARN = 'warn'
}

export default UserLogLevel
