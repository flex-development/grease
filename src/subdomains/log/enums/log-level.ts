/**
 * @file Enums - LogLevel
 * @module grease/log/enums/LogLevel
 */

import type { Opaque } from '@flex-development/tutils'

/**
 * Log levels.
 *
 * @enum {number}
 */
enum LogLevel {
  DEBUG = 4,
  ERROR = 0,
  FATAL = -1,
  INFO = 3,
  LOG = 2,
  SILENT = <Opaque<number, '-INFINITY'>>Number.NEGATIVE_INFINITY,
  TRACE = 5,
  VERBOSE = <Opaque<number, 'INFINITY'>>Number.POSITIVE_INFINITY,
  WARN = 1
}

export default LogLevel
