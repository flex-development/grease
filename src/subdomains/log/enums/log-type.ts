/**
 * @file Enums - LogType
 * @module grease/log/enums/LogType
 */

/**
 * Log types.
 *
 * @enum {Lowercase<string>}
 */
enum LogType {
  DEBUG = 'debug',
  ERROR = 'error',
  FAIL = 'fail',
  FATAL = 'fatal',
  INFO = 'info',
  LOG = 'log',
  START = 'start',
  SUCCESS = 'success',
  TRACE = 'trace',
  VERBOSE = 'verbose',
  WARN = 'warn'
}

export default LogType
