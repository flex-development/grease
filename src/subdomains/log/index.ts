/**
 * @file Entry Point - log
 * @module grease/log
 */

export * from './enums'
export type * from './interfaces'
export { default as LogModule } from './log.module'
export * from './models'
export * from './options'
export * from './providers'
export * from './reporters'
export type * from './types'
