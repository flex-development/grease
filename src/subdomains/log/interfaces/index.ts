/**
 * @file Entry Point - Interfaces
 * @module grease/log/interfaces
 */

export type { default as InputLog } from './input-log.interface'
export type { default as ILogger } from './logger.interface'
