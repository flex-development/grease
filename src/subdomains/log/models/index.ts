/**
 * @file Entry Point - Models
 * @module grease/log/models
 */

export { default as LogObject } from './log-object.model'
