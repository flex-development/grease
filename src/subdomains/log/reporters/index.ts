/**
 * @file Entry Point - Reporters
 * @module grease/log/reporters
 */

export { default as FancyReporter } from './fancy.reporter'
export { default as Reporter } from './reporter.abstract'
