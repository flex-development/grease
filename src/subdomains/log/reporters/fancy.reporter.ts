/**
 * @file Reporters - FancyReporter
 * @module grease/log/reporters/FancyReporter
 */

import { LogLevel, LogType } from '#src/log/enums'
import { LogObject } from '#src/log/models'
import type { Color } from '#src/log/types'
import pathe from '@flex-development/pathe'
import {
  fallback,
  ifelse,
  join,
  select,
  sift,
  split,
  template,
  trim,
  uppercase
} from '@flex-development/tutils'
import isUnicodeSupported from 'is-unicode-supported'
import util from 'node:util'
import sw from 'string-width'
import wrap from 'word-wrap'
import Reporter from './reporter.abstract'

/**
 * Fancy log reporter.
 *
 * @see {@linkcode Reporter}
 *
 * @class
 * @extends {Reporter}
 */
class FancyReporter extends Reporter {
  /**
   * Format an icon.
   *
   * @see {@linkcode LogObject}
   *
   * @protected
   *
   * @param {LogObject} log - Log object
   * @return {string} Formatted icon
   */
  protected formatIcon(log: LogObject): string {
    /**
     * Log object icon.
     *
     * @const {string} icon
     */
    const icon: string = this.icon(log)

    // exit early if no icon to format
    if (!icon) return icon

    /**
     * Background or icon color.
     *
     * @var {Color} color
     */
    let color: Color

    // get color
    switch (<Exclude<LogType, LogType.LOG>>log.type) {
      case LogType.DEBUG:
      case LogType.TRACE:
      case LogType.VERBOSE:
        color = 'gray'
        break
      case LogType.ERROR:
      case LogType.FATAL:
        color = 'bgRed'
        break
      case LogType.FAIL:
        color = 'red'
        break
      case LogType.INFO:
        color = 'cyan'
        break
      case LogType.START:
        color = 'magenta'
        break
      case LogType.SUCCESS:
        color = 'green'
        break
      case LogType.WARN:
        color = 'yellow'
        break
    }

    return this.logger.colors[color](ifelse(
      log.level <= LogLevel.ERROR && log.type !== LogType.FAIL,
      template(' {type} ', { type: uppercase(log.type) }),
      icon
    ))
  }

  /**
   * Format a log object.
   *
   * @see {@linkcode LogObject}
   *
   * @protected
   *
   * @param {LogObject} log - Log object
   * @return {string} Formatted log object
   */
  protected formatLogObject(log: LogObject): string {
    // recursively format log data
    if (log.message instanceof LogObject) {
      return this.formatLogObject(log.message)
    }

    /**
     * Text appended to formatted log object.
     *
     * @const {string} header
     */
    const header: string = join(sift([
      this.formatTag(log),
      this.formatIcon(log)
    ]), ' ')

    /**
     * Indentation size.
     *
     * @const {string} indent
     */
    const indent: string = ' '.repeat(sw(header))

    /**
     * Formatted log message lines.
     *
     * @const {string[]} lines
     */
    const lines: string[] = split(util.formatWithOptions({
      breakLength: process.stdout.columns,
      colors: this.logger.options.color
    }, log.message, ...log.args), '\n')

    return template('{newline}{header} {lines}{trace}', {
      header,
      lines: join(select(lines, null, (line: string, i: number): string => {
        return ifelse(i, wrap(line, { indent, width: line.length }), line)
      }), '\n'),
      newline: ifelse(log.level <= LogLevel.ERROR && log.error, '\n', ''),
      trace: this.formatTrace(log)
    })
  }

  /**
   * Format a logger tag.
   *
   * @see {@linkcode LogObject}
   *
   * @protected
   *
   * @param {LogObject} log - Log object
   * @return {string} Formatted tag
   */
  protected formatTag(log: LogObject): string {
    return ifelse(log.tag, this.logger.colors.gray(template('[{tag}]', {
      tag: log.tag
    })), '')
  }

  /**
   * Format a stack trace.
   *
   * @see {@linkcode LogObject}
   *
   * @protected
   *
   * @param {LogObject} log - Log object
   * @return {string} Formatted stack trace
   */
  protected formatTrace(log: LogObject): string {
    /**
     * Stack trace.
     *
     * @var {string} stack
     */
    let stack: string

    // get stack trace
    switch (true) {
      case !!log.error?.stack:
        stack = log.error!.stack!
        break
      default:
        return ''
    }

    return template('\n\n{trace}\n', {
      trace: join(select(split(stack, '\n').splice(1), null, line => {
        return ' '.repeat(sw(this.formatTag(log)) + 3) + trim(line)
          .replace(/(?<= \()file:\/\//, '')
          .replace(process.cwd() + pathe.sep, '')
          .replace(/^at+/, this.logger.colors.gray)
          .replace(/(?<=\().+(?=\))/, this.logger.colors.cyan)
      }), '\n')
    })
  }

  /**
   * Get a log object icon.
   *
   * @see {@linkcode LogObject}
   *
   * @protected
   *
   * @param {LogObject} log - Log object
   * @return {string} Log object icon
   */
  protected icon(log: LogObject): string {
    /**
     * Log object icon.
     *
     * @var {string} icon
     */
    let icon: string

    // get icon
    switch (log.type) {
      case LogType.DEBUG:
        icon = fallback('D', '⚙', isUnicodeSupported)
        break
      case LogType.ERROR:
      case LogType.FAIL:
      case LogType.FATAL:
        icon = fallback('×', '✖', isUnicodeSupported)
        break
      case LogType.INFO:
        icon = fallback('i', 'ℹ', isUnicodeSupported)
        break
      case LogType.LOG:
        icon = ''
        break
      case LogType.START:
        icon = fallback('o', '◐', isUnicodeSupported)
        break
      case LogType.SUCCESS:
        icon = fallback('√', '✔', isUnicodeSupported)
        break
      case LogType.TRACE:
        icon = fallback('→', '→', isUnicodeSupported)
        break
      case LogType.VERBOSE:
        icon = fallback('>', '❯', isUnicodeSupported)
        break
      case LogType.WARN:
        icon = fallback('‼', '⚠', isUnicodeSupported)
        break
    }

    return icon
  }

  /**
   * Write a log.
   *
   * @see {@linkcode LogObject}
   *
   * @public
   *
   * @param {LogObject} log - Log object to write
   * @return {void} Nothing when complete
   */
  public write(log: LogObject): void {
    /**
     * Writer data.
     *
     * @const {string} data
     */
    const data: string = this.formatLogObject(log) + '\n'

    /**
     * Writer name suffix.
     *
     * @const {'err' | 'out'} writer
     */
    const writer: 'err' | 'out' = log.level < LogLevel.LOG ? 'err' : 'out'

    // write log to stderr or stdout based on log level
    return void process[`std${writer}`].write(data)
  }
}

export default FancyReporter
