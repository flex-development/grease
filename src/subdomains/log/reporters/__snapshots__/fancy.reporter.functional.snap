// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`functional:log/reporters/FancyReporter > #write > LogType.DEBUG > should write log 1`] = `
"
[90m[grease:config][39m [90m⚙[39m invalid extension .yml
"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.ERROR > should write log 1`] = `
"
[90m[grease:bump][39m [41m ERROR [49m Invalid Version: 3.0

                [90mat[39m src/subdomains/log/reporters/__tests__/fancy.reporter.functional.spec.ts:34:18
                [90mat[39m file://node_modules/@vitest/runner/dist/index.js:361:43
                [90mat[39m runWithSuite ([36mnode_modules/@vitest/runner/dist/index.js:50:9[39m)
                [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:361:13[39m)
                [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:364:57[39m)
                [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:364:49[39m)
                [90mat[39m collectTests ([36mnode_modules/@vitest/runner/dist/index.js:512:28[39m)
                [90mat[39m startTests ([36mnode_modules/@vitest/runner/dist/index.js:840:17[39m)
                [90mat[39m file://node_modules/vitest/dist/entry.js:103:7
                [90mat[39m withEnv ([36mnode_modules/vitest/dist/entry.js:73:5[39m)

"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.FAIL > should write log 1`] = `
"
[90m[grease:config][39m [31m✖[39m Could not resolve \\"logger\\"

                  [90mat[39m src/subdomains/log/reporters/__tests__/fancy.reporter.functional.spec.ts:38:18
                  [90mat[39m file://node_modules/@vitest/runner/dist/index.js:361:43
                  [90mat[39m runWithSuite ([36mnode_modules/@vitest/runner/dist/index.js:50:9[39m)
                  [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:361:13[39m)
                  [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:364:57[39m)
                  [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:364:49[39m)
                  [90mat[39m collectTests ([36mnode_modules/@vitest/runner/dist/index.js:512:28[39m)
                  [90mat[39m startTests ([36mnode_modules/@vitest/runner/dist/index.js:840:17[39m)
                  [90mat[39m file://node_modules/vitest/dist/entry.js:103:7
                  [90mat[39m withEnv ([36mnode_modules/vitest/dist/entry.js:73:5[39m)

"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.FATAL > should write log 1`] = `
"
[90m[grease:git][39m [41m FATAL [49m not a git repository

               [90mat[39m src/subdomains/log/reporters/__tests__/fancy.reporter.functional.spec.ts:42:18
               [90mat[39m file://node_modules/@vitest/runner/dist/index.js:361:43
               [90mat[39m runWithSuite ([36mnode_modules/@vitest/runner/dist/index.js:50:9[39m)
               [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:361:13[39m)
               [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:364:57[39m)
               [90mat[39m Object.collect ([36mnode_modules/@vitest/runner/dist/index.js:364:49[39m)
               [90mat[39m collectTests ([36mnode_modules/@vitest/runner/dist/index.js:512:28[39m)
               [90mat[39m startTests ([36mnode_modules/@vitest/runner/dist/index.js:840:17[39m)
               [90mat[39m file://node_modules/vitest/dist/entry.js:103:7
               [90mat[39m withEnv ([36mnode_modules/vitest/dist/entry.js:73:5[39m)

"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.INFO > should write log 1`] = `
"
[90m[grease:git][39m [36mℹ[39m using git version 2.39.0 (/usr/local/bin/git)
"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.LOG > should write log 1`] = `
"
log message
"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.START > should write log 1`] = `
"
[90m[grease:tag][39m [35m◐[39m verifying gpg signature
"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.SUCCESS > should write log 1`] = `
"
[90m[grease:tag][39m [32m✔[39m object 067c93cc68575a652d8609585bd64478c38dccfe
               type commit
               tag grease@3.0.0-dev.1
               tagger Lexus Drumgold <unicornware@flexdevelopment.llc> 1696978824 -0400
               
               release: grease@3.0.0-dev.1
"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.TRACE > should write log 1`] = `
"
[90m[grease:git][39m [90m→[39m cannot describe '067c93cc68575a652d8609585bd64478c38dccfe'

               [90mat[39m src/subdomains/log/reporters/__tests__/fancy.reporter.functional.spec.ts:84:15
               [90mat[39m file://node_modules/@vitest/runner/dist/index.js:58:26
               [90mat[39m file://node_modules/@vitest/runner/dist/index.js:582:59
               [90mat[39m Array.map ([36m<anonymous>[39m)
               [90mat[39m callSuiteHook ([36mnode_modules/@vitest/runner/dist/index.js:582:47[39m)
               [90mat[39m runSuite ([36mnode_modules/@vitest/runner/dist/index.js:765:33[39m)

"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.VERBOSE > should write log 1`] = `
"
[90m[grease:config][39m [90m❯[39m Read 19 entries for directory \\"/\\"
"
`;

exports[`functional:log/reporters/FancyReporter > #write > LogType.WARN > should write log 1`] = `
"
[90m[grease:config][39m [33m⚠[39m This assignment will throw because \\"foo\\" is a constant [assign-to-constant]
"
`;
